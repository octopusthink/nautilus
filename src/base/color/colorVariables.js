// ----------------------------------------
// Color Variables
//
// https://ether.thescenery.co/color/
//
//

import colors from './colors';

export const semanticColors = {
  colorSuccess: colors.greenDark,
  colorDanger: colors.redDark,
};

export const typographicColors = {
  typeLight: colors.gray30,
  typeMedium: colors.gray50,
  typeDark: colors.black,
  typeAccent: colors.blue,
  typeDanger: colors.red,
  typeSuccess: colors.green,
};

export const buttonColors = {
  // Primary
  buttonPrimaryBackground: colors.blue,
  buttonPrimaryBackgroundHover: colors.blueDark1,
  buttonPrimaryBorder: colors.blue,
  buttonPrimaryText: colors.white,

  // Secondary
  buttonSecondaryBackground: colors.white,
  buttonSecondaryBackgroundHover: colors.gray15,
  buttonSecondaryBorder: colors.gray40,
  buttonSecondaryColor: colors.black,

  // Danger
  buttonDangerBackground: semanticColors.colorDanger,
  buttonDangerBackgroundHover: colors.redDark1,
  buttonDangerBorder: semanticColors.colorDanger,
  buttonDangerText: colors.white,

  // Success
  buttonSuccessBackground: semanticColors.colorSuccess,
  buttonSuccessBackgroundHover: colors.greenDark1,
  buttonSuccessBorder: semanticColors.colorSuccess,
  buttonSuccessText: colors.white,
};

const colorVariables = {
  // Semantic Colors
  semantic: {
    ...semanticColors,
  },

  // Typographic Colors
  typographic: {
    ...typographicColors,
  },

  // Button Colors
  buttons: {
    ...buttonColors,
  },
};

export default colorVariables;
