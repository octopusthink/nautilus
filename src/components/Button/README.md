Use a button any time the user should complete an action.

```js
const onClick = (event) => {
	window.alert('ðŸš¨ Alarm sounded! ðŸš¨');
};
<Button onClick={onClick}>Sound the alarm!</Button>;
```

## Best practises

- Keep labels as **short** as possible.
- Labels should use **sentence case**, not title case. (_Press me_, not _Press Me_.)
- Use prominence to **highlight the most important action** a user can take.
- **Use icons** to convey additional meaning, when appropriate.

## Properties

### Prominence

**Options:** primary, default, minimal **Changes:** styling (border and background)

Buttons have three levels of prominence. The majority of buttons should use the default styling. Use a primary button to draw attention to the most important action on the page. Don't use more than one primary button on a page. Minimal buttons are good when you don't want to draw too much attention to a button.

```js
<React.Fragment>
	<Button primary>Sign up now</Button>
	<Button>Sign up sometime</Button>
	<Button minimal>Sign up eventually</Button>
</React.Fragment>
```

### Behaviour

**Options:** navigation, action **Changes:** Adds an arrow to indicate semantics.

Sometimes, you need a button that's actually a link. We got you!

In these cases, we can use the same element, but apply ever-so-slightly-different styling to indicate the behaviour isn't exactly the same.

```js
<Button href="https://twitter.com/KittenBreak" navigation target="_blank">
	Take a break
</Button>

<Button href="https://twitter.com/KittenBreak" navigation primary>
	Take a break
</Button>

<Button href="https://twitter.com/KittenBreak" navigation minimal>
	Take a break
</Button>
```

### Intent

**Options:** success, warning, danger, none **Changes:** colour

Sometimes, you need a button to communicate the choices it's offering up more clearly. In these cases, you may want to declare the _intent_ of the button. This uses colour (and an icon, to ensure colour isn't the sole indicator of meaning)

### State

**Options:** active, hover, focussed, disabled, default **Changes:** styling

```js
<React.Fragment>
	<Button primary>Active</Button>
	<Button>Hover</Button>
	<Button minimal>Focus</Button>
</React.Fragment>
```

```js
<React.Fragment>
	<Button primary disabled>
		Disabled primary
	</Button>
	<Button disabled>Disabled</Button>
	<Button minimal disabled>
		Minimal disabled
	</Button>
</React.Fragment>
```

What's my button up to? State will tell you!

## Future improvements

- [ ] Button can include icons + text, just text, or just icons
- [ ] Icon can go before or after text
- [ ] Include icons in button intent
- [ ] Split buttons?
- [ ] Busy button (maybe just reuse active state)
- [ ] Include an icon in active styling (spinner?)
