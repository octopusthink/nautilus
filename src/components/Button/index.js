import { css } from '@emotion/core';
import invariant from 'invariant';
import PropTypes from 'prop-types';
import React from 'react';

export const Button = ({
  children,
  minimal,
  navigation,
  primary,
  ...otherProps
}) => {
  invariant(
    (minimal === true && primary === true) === false,
    'Cannot use `minimal` and `primary` props together on a <Button> component.',
  );

  let Component = 'button';
  if (navigation === true) {
    Component = 'a';
  }

  return (
    // See: https://github.com/yannickcr/eslint-plugin-react/issues/1555
    // eslint-disable-next-line react/button-has-type
    <Component
      css={css`
        font-family: -apple-system;
        font-size: 17px;
        padding: 12px 32px;
        outline: none;
        text-decoration: none;
        border: 2px solid #c62e7f;
        color: #c62e7f;
        background: #fff;
        border-radius: 8px;
        margin: 0 16px 16px;
        line-height: 1.412;
        transition: box-shadow 200ms;

        &:active {
          border-color: #973367;
          color: #973367;
        }

        &:focus {
          box-shadow: 0 0 1px 4px #F5A3CE;
          outline: none;
        }

        &::-moz-focus-inner {
          border: 0;
        }

        &:hover {
          box-shadow: 0 2px 0 0 #973367;
        }

        ${primary &&
          css`
            background: #c62e7f;
            color: #fff;

            &:active {
              background: #973367;
              color: #fff;
            }
          `}

        ${minimal &&
          css`
            border-top: 0;
            border-left: 0;
            border-right: 0;
            border-radius: 0;
            padding-left: 0;
            padding-right: 0;
          `}

        ${navigation &&
          css`
            &::after {
              content: ' â†’';
              display: inline;
            }
          `}

          &:disabled {
            border-color: #9f9f9f;
            color: #9f9f9f;

            &:hover {
                box-shadow: 0 2px 0 0 #6C6C6C;
            }

            &:active {
              background: #6C6C6C;
            }

            &:focus {
              box-shadow: 0 0 1px 4px #D4D4D4;
            }

            &${primary &&
              css`
                background: #9f9f9f;
                color: #fff;
              `}

      `}
      {...otherProps}
    >
      {children}
    </Component>
  );
};

Button.defaultProps = {
  children: undefined,
  disabled: false,
  minimal: false,
  navigation: false,
  primary: false,
  type: 'button',
};

Button.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Disables the button; this applies a disabled style but **does not disable any event handlers for the button**. Your `onClick` handlers should check for `props.disabled` to modify their behaviour accordingly. */
  disabled: PropTypes.bool,
  /** Decrease the visual prominence of this button in the UI. */
  minimal: PropTypes.bool,
  /** Sometimes, you need a button that's actually a link. We got you! This outputs a `react-router-dom` `<Link>` tag that looks (and largely behaves) like a `<Button>`, but can used as navigation. Setting this to `true` enables `Link` properties; see: https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/api/Link.md. */
  navigation: PropTypes.bool,
  /** Increase the visual prominence of this button in the UI. */
  primary: PropTypes.bool,
  /** HTML `type` attribute for the button. Defaults to `"button"`. */
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
};

export default Button;
