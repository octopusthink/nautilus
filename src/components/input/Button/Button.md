### Primary Button
```js
<Button size="small">Small Button</Button>
```
```js
<Button>Medium Button</Button>
```
```js
<Button size="large">Large Button</Button>
```

### Secondary Button
```js
<Button mode="secondary">Button</Button>
```

### Danger Button
```js
<Button mode="danger">Button</Button>
```

### Success Button
```js
<Button mode="success">Button</Button>
```
