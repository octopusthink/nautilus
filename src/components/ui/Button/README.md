A button indicates that a user can complete an action or make a choice, using a single click or tap. They are used for submitting forms, navigating to new pages, as well as toolbars and utilities.

```jsx
const onClick = (event) => {
  window.alert('ðŸš¨ Alarm sounded! ðŸš¨');
};
<Button onClick={onClick}>Sound the alarm!</Button>;
```

## Best practises

### Labels

- Keep labels as **short** and direct as possible. Omit articles and unnecessary words wherever possible.
- Use an **active verb + noun** form. (_Reserve table_, _Delete page_, etc.) In some cases, a verb alone may suffice. (_Save_, _Okay_, _Cancel_.)
- Labels should use **sentence case**, not title case. (_Press me_, not _Press Me_.)
- **Use icons** to convey additional meaning, when appropriate. For important actions, including both text and an icon is a good way to reinforce meaning.

### Prominence

- Only use a single **primary** button per screen.
- Use prominence to **highlight the most important action** a user can take.
- When using two or more buttons together, try to use different prominence levels to guide users to the most likely actions.

### Placement

- When buttons are used in a form, they should **always be placed after inputs**, both visually and in the DOM order.
- When using two or more buttons together, don't stack them unless there isn't enough space to align them side-by-side.

## Properties

### Prominence

**Options:** primary, default, minimal

**Changes:** styling (border and background)

Buttons have three levels of prominence. The majority of buttons should use the **default styling**.

Use a primary button to draw attention to the most important action on the page. Don't use more than one primary button on a page.

Minimal buttons are good when you don't want to draw too much attention to a button. Use them for secondary actions.

```jsx
<Button primary>Sign up now</Button>
<Button>Sign up sometime</Button>
<Button minimal>Sign up eventually</Button>
```

### Behaviour

**Options:** navigation, action

**Changes:** Adds an arrow to indicate semantics.

Sometimes, you need a button that's actually a link. (Maybe you want a call-to-action, or a link that's more prominent than a plain-text link.) We got you!

Wait, what's the difference, you ask?

A **button** completes an action, like deleting an item, submitting a form, or changing part of the UI. A **link** navigates to a new page or site, refreshing the page.

A link loses focus and means the user needs to reorient themselves, but a button doesn't.

This is an important different for accessibility needs, but as single page web apps become more common, the difference often starts to blur. In order to keep things flexible, a Button can also act as a link. We add an arrow to the text to indicate that the behaviour won't be exactly the same, and we use an `<a>` tag so screenreaders know what to expect as well.

```jsx
<Button navigation to="https://twitter.com/KittenBreak" useHref>
  Take a break
</Button>

<Button navigation primary to="https://twitter.com/KittenBreak" useHref>
  Take a break
</Button>

<Button navigation minimal to="https://twitter.com/KittenBreak" useHref>
  Take a break
</Button>
```

### Intent

**Options:** success, warning, danger, none

**Changes:** colour

Sometimes, you need a button to communicate the choices it's offering up more clearly. In these cases, you may want to declare the _intent_ of the button. This uses colour and an icon to indicate the messaging.

Use a danger button any time you have a destructive action, like deleting something.

```jsx
<Button primary success>
  Nobel Peace Prize
</Button>
<Button success>Got a new job</Button>
<Button minimal success>
  Woke up on time
</Button>
```

```jsx
<Button primary warning>
  Trump 4 prez
</Button>

<Button warning>Look out!</Button>

<Button minimal warning>
  Pothole in the road
</Button>
```

```jsx
<Button primary danger>
  Scorpion in your swimsuit
</Button>

<Button danger>Snakes on a plane</Button>

<Button minimal danger>
  Spider in the kitchen
</Button>
```

### State

**Options:** active, hover, focussed, disabled, default

**Changes:** styling

```jsx
<Button primary>Active</Button>
<Button>Hover</Button>
<Button minimal>Focus</Button>
```

```jsx
<Button primary disabled>
  Disabled primary
</Button>
<Button disabled>Disabled</Button>
<Button minimal disabled>
  Minimal disabled
</Button>
```

What's my button up to? State will tell you!
