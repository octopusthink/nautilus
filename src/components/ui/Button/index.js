import { css } from '@emotion/core';
import invariant from 'invariant';
import PropTypes from 'prop-types';
import React from 'react';

// Import the non-styled Link so we don't have to overwrite Link styles if
// the `navigation` prop is used by a button.
import Link from 'components/ui/Link';
import Icon from 'components/ui/Icon';
import { interfaceUI, toUnits } from 'styles';
import { useTheme } from 'themes';

export const ComponentClassName = 'Nautilus-Button';

export const qualityControl = (props) => {
  const { minimal, primary, success, warning, danger } = props;

  invariant(
    [minimal, primary].filter((prop) => prop).length <= 1,
    '<Button> should not be both `minimal` and `primary`, so which is it?',
  );

  invariant(
    [danger, success, warning].filter((prop) => prop).length <= 1,
    '<Button> should only use one of `danger`, `warning`, or `success`. Pick a lane!',
  );
};

export const Button = (props) => {
  const {
    __iconId,
    children,
    danger,
    disabled,
    href,
    linkProps,
    leadingIcon,
    minimal,
    navigation,
    noMargin,
    primary,
    stackedIcon,
    success,
    trailingIcon,
    unstyled,
    warning,
    ...otherProps
  } = props;

  qualityControl(props);
  const theme = useTheme();

  let Component = 'button';
  let linkPropsToUse;
  if (navigation === true) {
    Component = Link;
    // Set properties that only a Link component should use.
    otherProps.href = href;
    // Don't include any link styles.
    otherProps.unstyled = true;
    // Unset certain button-specific props.
    otherProps.type = undefined;
    linkPropsToUse = { ...linkProps };
  }

  let iconName;
  if (success === true) {
    iconName = 'check-circle';
  }
  if (warning === true) {
    iconName = 'alert-circle';
  }
  if (danger === true) {
    iconName = 'x-circle';
  }

  // Set the styles for this button.
  let currentButtonColor = theme.colors.buttons.default;
  let currentButtonColorDark = theme.colors.buttons.defaultDark;
  let currentButtonColorLight = theme.colors.buttons.defaultLight;

  if (disabled === true) {
    currentButtonColor = theme.colors.state.disabled;
    currentButtonColorDark = theme.colors.state.disabledDark;
    currentButtonColorLight = theme.colors.state.disabledLight;
  }

  if (success === true) {
    currentButtonColor = theme.colors.intent.success;
    currentButtonColorDark = theme.colors.intent.successDark;
    currentButtonColorLight = theme.colors.intent.successLight;
  }

  if (warning === true) {
    currentButtonColor = theme.colors.intent.warning;
    currentButtonColorDark = theme.colors.intent.warningDark;
    currentButtonColorLight = theme.colors.intent.warningLight;
  }

  if (danger === true) {
    currentButtonColor = theme.colors.intent.danger;
    currentButtonColorDark = theme.colors.intent.dangerDark;
    currentButtonColorLight = theme.colors.intent.dangerLight;
  }

  return (
    // See: https://github.com/yannickcr/eslint-plugin-react/issues/1555
    // eslint-disable-next-line react/button-has-type
    <Component
      css={
        unstyled
          ? undefined
          : css`
        ${interfaceUI.medium(theme)}
        align-items: center;
        background: ${theme.colors.buttons.neutral};
        border: 2px solid ${currentButtonColor};
        border-radius: 8px;
        color: ${currentButtonColor};
        cursor: pointer;
        display: flex;
        margin: 0;
        outline: none;
        padding: ${toUnits(theme.spacing.padding.medium)} ${toUnits(theme.spacing.padding.large)};
        position: relative;
        text-align: center;
        text-decoration: none;
        top: 0;
        transition: all 200ms ease-in-out;
        transition: top 0ms ease-out;
        width: max-content;

        &::-moz-focus-inner {
          border: 0;
        }

        &:active {
          border-color: ${currentButtonColorDark};
          color: ${currentButtonColorDark};
          top: 2px;
        }

        &:focus {
          box-shadow: 0 0 1px 4px ${currentButtonColorLight};
          outline: none;
        }

        /* Set padding based on whether we have leading or trailing icons. */
        ${leadingIcon &&
          css`
            padding-left: ${toUnits(theme.spacing.padding.medium)};
          `}

        ${trailingIcon &&
          css`
            padding-right: ${toUnits(theme.spacing.padding.medium)};
          `}

        ${stackedIcon &&
          css`
            flex-direction: column;
          `}

        /* Set external margins */
        ${!noMargin &&
          css`
            margin: 0 ${toUnits(theme.spacing.margin.xxSmall)}
              ${toUnits(theme.spacing.margin.xSmall)};
          `}

        /* Primary styles */
        ${primary &&
          css`
            background: ${currentButtonColor};
            color: ${theme.colors.buttons.neutral};

            &:hover {
              background-color: ${currentButtonColorDark};
            }

            &:active {
              background: ${currentButtonColorDark};
              color: ${theme.colors.buttons.neutral};
            }
          `}

        /* Minimal styles */
        ${minimal &&
          css`
            background: transparent;
            border: 0;
            padding-left: 0;
            padding-right: 0;
            position: relative;

            &::before {
              content: '';
              position: absolute;
              bottom: 0;
              width: 0;
              border-bottom: 2px solid ${currentButtonColorDark};
              transition: all 200ms ease-in-out;
            }

            &:hover::before {
              width: 100%;
            }
          `}

        ${navigation &&
          css`
            &::after {
              content: ' â†’';
              display: inline;
              padding-right: ${toUnits(theme.spacing.padding.xSmall)};
              transition: all 200ms ease-in-out;
            }

            &:hover::after {
              margin-left: ${toUnits(theme.spacing.padding.xSmall)};
              padding-right: 0;
            }
          `}

        &:hover {
          ${!minimal &&
            css`
              border-color: ${currentButtonColorDark};
              box-shadow: 0 2px 4px rgba(0, 0, 0, 0.25);
            `}

          ${!primary &&
            css`
              color: ${currentButtonColorDark};
            `}
        }
      `
      }
      disabled={!navigation ? disabled : undefined}
      {...linkPropsToUse}
      {...otherProps}
    >
      {iconName && (
        <Icon
          css={css`
            margin-right: ${toUnits(theme.spacing.padding.xSmall)};
          `}
          id={__iconId}
          name={iconName}
        />
      )}

      {leadingIcon && (
        <Icon
          css={css`
            margin: 0 ${toUnits(theme.spacing.padding.small)} 0 0;
          `}
          name={leadingIcon}
        />
      )}

      {stackedIcon && (
        <Icon
          css={css`
            margin: 0 0 ${toUnits(theme.spacing.padding.xSmall)} 0;
          `}
          name={stackedIcon}
          medium
        />
      )}

      {children}

      {trailingIcon && (
        <Icon
          css={css`
            margin: 0 0 0 ${toUnits(theme.spacing.padding.small)};
          `}
          name={trailingIcon}
        />
      )}
    </Component>
  );
};

Button.defaultProps = {
  __iconId: undefined,
  children: undefined,
  danger: false,
  disabled: false,
  href: undefined,
  leadingIcon: undefined,
  linkProps: undefined,
  minimal: false,
  navigation: false,
  noMargin: false,
  primary: false,
  stackedIcon: undefined,
  success: false,
  trailingIcon: undefined,
  type: 'button',
  unstyled: false,
  warning: false,
};

Button.propTypes = {
  /** @ignore ID prop for the Icon component; only used for testing. */
  __iconId: PropTypes.string,
  /** @ignore */
  children: PropTypes.node,
  /** Disables the button; this applies a disabled style but **does not disable any event handlers for the button**. Your `onClick`, `onTap`, etc. handlers should check for the `disabled` prop to modify their behaviour accordingly. */
  disabled: PropTypes.bool,
  /** Increase the visual prominence of this button in the UI. */
  primary: PropTypes.bool,
  /** Decrease the visual prominence of this button in the UI. */
  minimal: PropTypes.bool,
  /** Apply semantic styling to indicate success or positive intent. */
  success: PropTypes.bool,
  /** Apply semantic styling to indicate a warning. */
  warning: PropTypes.bool,
  /** Apply semantic styling to indicate danger or negative intent. */
  danger: PropTypes.bool,
  /** Used to link to a route that will be handled by Nautilus' `Link` component. */
  href: PropTypes.string,
  /** Show an icon inside the button, before the text. Passes a string to Icon's name prop. */
  leadingIcon: PropTypes.string,
  /** Props to pass to the underlying Nautilus `Link` component when `useNavigation` is `true`. */
  // eslint-disable-next-line react/forbid-prop-types
  linkProps: PropTypes.object,
  /** Outputs a Nautilus `<Link>` tag that looks (and largely behaves) like a `<Button>`, but can used as navigation. Setting this to `true` enables `Link` properties. */
  navigation: PropTypes.bool,
  /** Remove any outer margins from component. */
  noMargin: PropTypes.bool,
  /** Show an icon inside the button, on top of text. Passes a string to Icon's name prop. */
  stackedIcon: PropTypes.string,
  /** Show an icon inside the button, after the text. Passes a string to Icon's name prop. */
  trailingIcon: PropTypes.string,
  /** HTML `type` attribute for the button. Defaults to `"button"`. */
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
  /* @ignore Don't output any CSS styles. */
  unstyled: PropTypes.bool,
};

Button.displayName = 'Button';

export const { defaultProps, propTypes } = Button;

export default Button;
