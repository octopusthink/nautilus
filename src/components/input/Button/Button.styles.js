import styled, { css } from 'styled-components';

import { buttonColors } from '../../../base/color/colorVariables';
import fontSizes from '../../../base/typography/fontSizes';

export const buttonModes = {
  primary: css`
    background-color: ${buttonColors.buttonPrimaryBackground};
    border: 1px solid ${buttonColors.buttonPrimaryBorder};
    color: ${buttonColors.buttonPrimaryText};

    &:hover {
      background-color: ${buttonColors.buttonPrimaryBackgroundHover};
    }
  `,
  secondary: css`
    background-color: ${buttonColors.buttonSecondaryBackground};
    border: 1px solid ${buttonColors.buttonSecondaryBorder};
    color: ${buttonColors.buttonSecondaryText};

    &:hover {
      background-color: ${buttonColors.buttonSecondaryBackgroundHover};
    }
  `,
  danger: css`
    background-color: ${buttonColors.buttonDangerBackground};
    border: 1px solid ${buttonColors.buttonDangerBorder};
    color: ${buttonColors.buttonDangerText};

    &:hover {
      background-color: ${buttonColors.buttonDangerBackgroundHover};
    }
  `,
  success: css`
    background-color: ${buttonColors.buttonSuccessBackground};
    border: 1px solid ${buttonColors.buttonSuccessBorder};
    color: ${buttonColors.buttonSuccessText};

    &:hover {
      background-color: ${buttonColors.buttonSuccessBackgroundHover};
    }
  `,
};

export const buttonSizes = {
  small: css`
    ${fontSizes.size0};
    padding: 8px 11px;
  `,
  medium: css`
    ${fontSizes.size1};
    padding: 10px 12px;
  `,
  large: css`
    ${fontSizes.size2};
    padding: 12px 15px;
  `,
};

export const StyledButton = styled.button`
  ${p => buttonModes[p.mode]};
  ${p => buttonSizes[p.size]};
  border-radius: 3px;
  cursor: pointer;
`;
