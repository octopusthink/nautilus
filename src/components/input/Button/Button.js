import PropTypes from 'prop-types';
import React from 'react';

import { StyledButton, buttonModes, buttonSizes } from './Button.styles';

export const Button = ({ children, mode, onClick, size }) => (
  <StyledButton
    onClick={onClick}
    mode={mode}
    size={size}
  >
    {children}
  </StyledButton>
);

export const ButtonModes = Object.keys(buttonModes);
export const ButtonSizes = Object.keys(buttonSizes);

Button.propTypes = {
  /** Content to be rendered inside a Button */
  children: PropTypes.node,
  /** Mode of the button. Options are 'primary', 'secondary', 'danger', 'success' */
  mode: PropTypes.oneOf(ButtonModes),
  /** Function to run on click */
  onClick: PropTypes.func,
  /** Size of the button. Options are 'small', 'medium', 'large' */
  size: PropTypes.oneOf(ButtonSizes),
};

Button.defaultProps = {
  mode: 'primary',
  onClick: () => {},
  size: 'medium',
};
