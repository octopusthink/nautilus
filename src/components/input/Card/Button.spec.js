import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';

import { Button, ButtonModes, ButtonSizes } from './Button';
import { buttonColors } from '../../../base/color/colorVariables';

describe('Button Component', () => {
  context('when nothing specified', () => {
    it('should render without issue', () => {
      const subject = renderer.create(<Button>Hello</Button>).toJSON();
      expect(subject).toMatchSnapshot();
      expect(subject).toHaveStyleRule('background-color', buttonColors.buttonPrimaryBackground);
    });
  });

  ButtonModes.forEach((mode) => {
    context(`when @mode specified as ${mode}`, () => {
      it('renders with correct @mode value', () => {
        const subject = mount(<Button mode={mode}>Hello</Button>);
        expect(subject.prop('mode')).toBe(mode);
      });

      // Simple background color value and variable name test
      it(`renders with correct colors for ${mode}`, () => {
        const subject = renderer.create(<Button mode={mode}>Hello</Button>).toJSON();
        const capitalMode = `${mode.charAt(0).toUpperCase()}${mode.substr(1)}`;
        const buttonBackground = buttonColors[`button${capitalMode}Background`];
        const buttonBackgroundHover = buttonColors[`button${capitalMode}BackgroundHover`];

        expect(subject).toHaveStyleRule('background-color', buttonBackground);

        expect(subject).toHaveStyleRule('background-color', buttonBackgroundHover, {
          modifier: ':hover',
        });
      });
    });
  });

  ButtonSizes.forEach((size) => {
    context(`when @size specified as ${size}`, () => {
      it('renders with correct @size value', () => {
        const subject = mount(<Button size={size}>Hello</Button>);
        expect(subject.prop('size')).toBe(size);
      });
    });
  });
});
