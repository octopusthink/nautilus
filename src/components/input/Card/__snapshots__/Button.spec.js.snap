// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component when nothing specified should render without issue 1`] = `
.c0 {
  background-color: #519CB0;
  border: 1px solid #519CB0;
  color: #FFFFFF;
  font-size: 2.04ex;
  padding: 10px 12px;
  border-radius: 3px;
  cursor: pointer;
}

.c0:hover {
  background-color: #27768B;
}

@media (min-width:768px) {
  .c0 {
    font-size: 2.08ex;
  }
}

<button
  className="c0"
  mode="primary"
  onClick={[Function]}
  size="medium"
>
  Hello
</button>
`;
